listenbrainz/webserver/views/atom.py

```diff
@@ -0,0 +1,592 @@
+    _entity_descriptions = {
+        "artists": "Top Artists",
+        "release_groups": "Top Albums",
+        "recordings": "Top Tracks",
+    }
+    _stats_range_descriptions = {
+        StatisticsRange.this_week.value: "This Week's",
+        StatisticsRange.this_month.value: "This Month's",
+        StatisticsRange.this_year.value: "This Year's",
+        StatisticsRange.week.value: "Weekly",
+        StatisticsRange.month.value: "Monthly",
+        StatisticsRange.quarter.value: "Quarterly",
+        StatisticsRange.half_yearly.value: "Half Yearly",
+        StatisticsRange.year.value: "Yearly",
+        StatisticsRange.all_time.value: "All Time",
+    }
+    return f"{_stats_range_descriptions[stats_range]} {_entity_descriptions[entity]} for {user_name} - ListenBrainz"
+
+
+def _get_stats_entry_title(stats_range: str, timestamp: int) -> str:
+    """
+    A central place to generate feed entry titles for stats feeds.
+    """
+    dt = datetime.fromtimestamp(timestamp, tz=timezone.utc)
+
+    if stats_range in [StatisticsRange.this_week.value, StatisticsRange.week.value]:
+        return f"Week {dt.isocalendar()[1]}"
+    elif stats_range in [StatisticsRange.this_month.value, StatisticsRange.month.value]:
+        return dt.strftime("%B %Y")
+    elif stats_range in [StatisticsRange.this_year.value, StatisticsRange.year.value]:
+        return dt.strftime("%Y")
+    elif stats_range in [StatisticsRange.quarter.value]:
+        return f"Q{int((dt.month - 1) / 3) + 1} {dt.year}"
+    elif stats_range in [StatisticsRange.half_yearly.value]:
+        return f"H{int((dt.month - 1) / 6) + 1} {dt.year}"
+    elif stats_range in [StatisticsRange.all_time.value]:
+        return "All Time"
+    assert False, f"Invalid stats_range: {stats_range}"
+
+
+def _init_feed(id, title, self_url, alternate_url):
+    """
+    Initialize a feed with common attributes.
+    """
+    fg = FeedGenerator()
+    fg.id(id)
+    fg.title(title)
+    fg.author({"name": "ListenBrainz"})
+    fg.link(href=alternate_url, rel="alternate")
+    fg.link(href=self_url, rel="self")
+    fg.logo(_external_url_for("static", filename="img/listenbrainz_logo_icon.svg"))
+    fg.language("en")
+    return fg
+
+
+@atom_bp.route("/user/<user_name>/listens", methods=["GET"])
+@crossdomain
+@ratelimit()
+@api_listenstore_needed
+def get_listens(user_name):
+    """
+    Get listens feed for a user.
+
+    :param minutes: The time interval in minutes from current time to fetch listens for.
+                    For example, if minutes=60, listens from the last hour will be fetched. Default is 60.
+    :statuscode 200: The feed was successfully generated.
+    :statuscode 400: Bad request.
+    :statuscode 404: The user does not exist.
+    :resheader Content-Type: *application/atom+xml*
+    """
+    user = db_user.get_by_mb_id(db_conn, user_name)
+    if user is None:
+        return NotFound("User not found")
+
+    minutes = request.args.get("minutes", DEFAULT_MINUTES_OF_LISTENS)
```

Comment:
Not sure minutes is a good idea, probably just accept a `from_ts` and forward it to fetch_listens. Maybe also accept `to_ts` (default to now), and add a check that from_ts and to_ts are not more than a week apart or something.

Reply:
Ah right, makes sense in that case.

Reply:
So instead of accepting `from_ts` just calculate it from `to_ts` and `minutes`.

Reply:
That sounds good.

Reply:
Thanks! But I'm quite sure the supposed change here. For feeds, to_ts is always now(), which is in the code.

Reply:
Yes, to_ts does not need to change. 

But fetch_listens also accepts a `from_ts`, 

https://github.com/metabrainz/listenbrainz-server/blob/150d2bc59f381a25788892fe172bc2fea6afaed9/listenbrainz/listenstore/timescale_listenstore.py#L178

You should calculate `from_ts = to_ts - timedelta(minutes=minutes)` and then pass it to fetch_listens instead of passing `limit = minutes`.

Reply:
That's what I did at first. But it's very likely that we won't be able to fetch all listens within a time range with the default limit. Is it?

Reply:
If you want to use minutes that way, then just rename minutes argument to `limit`. But I think using `from_ts` with a default of past 24 hours would be a better option, performance wise.

Reply:
Ah right. Is there a way to only pass `from_ts` and `to_ts` to `fetch_listens` and get all the listens within that range? I see the `LIMIT :limit` in the query and thought that we always have to pass some reasonable limit number or we will be capped by default.